cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java

```diff
@@ -0,0 +1,203 @@
+                String key = getApiKey(ctx);
+                String user = authorized(ctx, key, routeRoles);
+                prepareContextWithUser(ctx, user,key);
+            }
+            handler.handle(ctx);
+        }
+        catch(CwmsAuthException ex) {
+            logger.log(Level.WARNING,"Unauthorized login attempt",ex);
+            HashMap<String,String> msg = new HashMap<>();
+            msg.put("message",ex.getMessage());
+            RadarError re = new RadarError("Unauthorized",msg,true);
+            ctx.status(ex.getAuthFailCode()).json(re);
+        }		
+
+	}
+
+    /**
+     * This isn't currently used from Tomcat so we don't have any issues with the 
+     * session; just use the connection itself.
+     * 
+     * NOTE: we will need to expand a bit in the future, but I think another JAPSIC or valve
+     * handler that sets up the principal for the CWMS Access Manager will likely
+     * have easier workflow.
+     * @param ctx
+     * @param user
+     * @param key the key
+     */
+    private void prepareContextWithUser(Context ctx, String user,String key) throws SQLException {
+        logger.info("Validated Api Key for user=" + user);
+
+        ConnectionPreparer keyPreparer = new ApiKeyUserPreparer(key);
+        ConnectionPreparer officePrepare = new SessionOfficePreparer(ctx.queryParam("office"));
+        DelegatingConnectionPreparer apiPreparer = new DelegatingConnectionPreparer(keyPreparer,officePrepare);
+
+        if (dataSource instanceof ConnectionPreparingDataSource) {
+            ConnectionPreparingDataSource cpDs = (ConnectionPreparingDataSource)    dataSource;
+            ConnectionPreparer existingPreparer = cpDs.getPreparer();
+
+            // Have it do our extra step last.
+            cpDs.setPreparer(new DelegatingConnectionPreparer(existingPreparer, apiPreparer));
+        } else {            
+            ctx.attribute(ApiServlet.DATA_SOURCE, new ConnectionPreparingDataSource(apiPreparer, dataSource));
+        }
+    }
+
+    private String authorized(Context ctx, String key, Set<RouteRole> routeRoles)
+    {
+        String retval = null;
+        String user = checkKey(key,ctx);
+
+        if (routeRoles == null || routeRoles.isEmpty()) {
+            retval = user;
+        } else {
+            Set<RouteRole> specifiedRoles = getRoles(user,ctx.queryParam("office"));
+            if (specifiedRoles.containsAll(routeRoles) ) {
+                retval = user;
+            } else {
+                throw new CwmsAuthException("Operation not authorized for user",403);
+            }
+        }
+
+        return retval;
+    }
+
+    private Set<RouteRole> getRoles(String user,String office) {
+        Set<RouteRole> roles = new HashSet<>();
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper(?)); cwms_env.set_session_office_id(?);end;");
```

Comment:
you pulled the other statements into constants, why not all of them?

Reply:
would you believe, laziness. But good to call that our. I will adjust.

Reply:
good call, managed to leave a hardcoded office in one instance.